File: src/main/scala/io/boson/json/jsonImpl/JosonExtractor.scala
package io.boson.json.jsonImpl

import java.io.{ByteArrayOutputStream, IOException}
import java.util.concurrent.CompletableFuture

import com.fasterxml.jackson.databind.module.SimpleModule
import com.fasterxml.jackson.databind.ObjectMapper
import de.undercouch.bson4jackson.BsonFactory
import io.boson.bson.bsonImpl.BosonImpl
import io.boson.bson.bsonPath.{Interpreter, Program, TinyLanguage}
import io.boson.bson.bsonValue
import io.boson.bson.bsonValue.{BsBoson, BsValue}
import io.boson.json.Joson
import io.boson.json.Joson.{JsonArraySerializer, JsonObjectSerializer}
import io.vertx.core.json.{JsonArray, JsonObject}


/**
  * Created by Ricardo Martins on 19/01/2018.
  */



class JosonExtractor[T](expression: String, extractFunction: java.util.function.Consumer[T]) extends Joson {
  /**
    * Apply this Joson to the String that arrives and at some point in the future complete
    * the future with the resulting String. In the case of an Extractor this will result in
    * the immutable String being returned unmodified.
    *
    * @param the Json string.
    * @return
    */
  private def callParse(boson: BosonImpl, expression: String): io.boson.bson.bsonValue.BsValue = {
    val parser = new TinyLanguage
    try {
      parser.parseAll(parser.program, expression) match {
        case parser.Success(r, _) =>
          new Interpreter(boson, r.asInstanceOf[Program]).run()
        case parser.Error(msg, _) => bsonValue.BsObject.toBson(msg)
        case parser.Failure(msg, _) => bsonValue.BsObject.toBson(msg)
      }
    } catch {
      case e: RuntimeException => bsonValue.BsObject.toBson(e.getMessage)
    }
  }

  override def go(jsonStr: String): CompletableFuture[String] = {
    val a: JsonObject = new JsonObject(jsonStr)

    val mapper: ObjectMapper = new ObjectMapper(new BsonFactory())
    val os = new ByteArrayOutputStream
    val module = new SimpleModule
    module.addSerializer(classOf[JsonObject],new JsonObjectSerializer)
    module.addSerializer(classOf[JsonArray], new JsonArraySerializer)
    mapper.registerModule(module)

    mapper.writeValue(os, a)

    val bsonByteEncoding: Array[Byte] = os.toByteArray
    os.flush()

    val future: CompletableFuture[String] =
      CompletableFuture.supplyAsync(() => {
        val boson: io.boson.bson.bsonImpl.BosonImpl = new BosonImpl(byteArray = Option(bsonByteEncoding))
        callParse(boson, expression) match {
          case (res: BsValue) =>
            extractFunction.accept(res.asInstanceOf[T])
          case _ => throw new RuntimeException("JosonExtractor -> go() default case!!!")
        }
        jsonStr
      })
    future
  }

  /**
    * Fuse one Joson to another. The joson that is 'this' should be executed first before the
    * joson that is the parameter in teh case of update/read conflicts.
    * the immutable String being returned unmodified.
    *
    * @param the Joson to fuse to.
    * @return the fused Joson
    */
  override def fuse(joson: Joson): Joson = ???
}

File: boson-java/src/main/java/io/zink/joson/JosonExtractor.java
package io.zink.joson;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import de.undercouch.bson4jackson.BsonFactory;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.zink.boson.bson.bsonImpl.BosonImpl;
import io.zink.boson.bson.bsonPath.Interpreter;
import io.zink.boson.bson.bsonPath.Program;
import io.zink.boson.bson.bsonPath.TinyLanguage;
import io.zink.boson.bson.bsonValue.BsException$;
import io.zink.boson.bson.bsonValue.BsObject$;
import io.zink.boson.bson.bsonValue.BsValue;
import io.zink.boson.bson.bsonValue.Writes$;
import io.zink.bosonInterface.Boson;
import io.zink.josonInterface.Joson;
import scala.Function1;
import scala.Option;
import scala.util.parsing.combinator.Parsers;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;
import java.util.function.Function;

public class JosonExtractor<T> implements Joson {


    private String expression;
    private Consumer<T> extractFunction;


    public JosonExtractor(String expression, Consumer<T> extractFunction) {
        this.expression = expression;
        this.extractFunction = extractFunction;
    }

    private Function<String, BsValue> writer = (str) -> BsException$.MODULE$.apply(str);

    private BsValue callParse(BosonImpl boson, String expression){
        TinyLanguage parser = new TinyLanguage();
        try{
            Parsers.ParseResult pr = parser.parseAll(parser.program(), expression);
            if(pr.successful()){
                Interpreter interpreter = new Interpreter<>(boson, (Program) pr.get(), Option.empty());
                return interpreter.run();
            }else{
                return BsObject$.MODULE$.toBson("Failure/Error parsing!", Writes$.MODULE$.apply1(writer));
            }
        }catch (RuntimeException e){
            return BsObject$.MODULE$.toBson(e.getMessage(), Writes$.MODULE$.apply1(writer));
        }
    };

    @Override
    public CompletableFuture<String> go(String jsonStr) {

        JsonObject a = new JsonObject(jsonStr);
        ObjectMapper mapper = new ObjectMapper(new BsonFactory());
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        SimpleModule module = new SimpleModule();
        module.addSerializer(JsonObject.class, new Joson.JsonObjectSerializer());
        module.addSerializer(JsonArray.class, new Joson.JsonArraySerializer());
        mapper.registerModule(module);
        byte[] bsonByteEncoding;
        CompletableFuture<String> future;
        try {
            mapper.writeValue(os, a);
            bsonByteEncoding = os.toByteArray();
            os.flush();
            future =
                    CompletableFuture.supplyAsync(() -> {
                        Option<byte[]> opt = Option.apply(bsonByteEncoding);
                        Option e = Option.empty();
                        BosonImpl boson = new BosonImpl(opt, e,e);
                        BsValue value = callParse(boson, expression);
                        extractFunction.accept((T)value);
                        return jsonStr;
                    });
        }catch (IOException e){
            System.out.println(e.getMessage());
            future = new CompletableFuture<>();
            future.complete(jsonStr);

        }
        return future;
    }

    @Override
    public Joson fuse(Joson joson) {
        return new JosonFuse(this, joson);
    }
}
